### PR TITLE
 github ID validation

### PR DESCRIPTION
**Pull request by w8896699**
https://app.zenhub.com/workspaces/platform-experience-5bb7c5ab4b5806bc2beb9d15/issues/bcdevops/developer-experience/1255

### CODE REVIEW

File: web/src/components/common/UI/GithubUserValidation/AdaptedGithubUserDisplay.tsx [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+}
+
+const User: React.FC<GithubUserInterface> = (props) => {
+  const { name, id, avatar } = props;
+
+  return (
+    <Flex flexDirection="row" justifyContent="space-evenly">
+      <Box
+        sx={{
+          width: '10%',
```
Comment (patricksimonian) [inline web/src/components/common/UI/GithubUserValidation/AdaptedGithubUserDisplay.tsx:27]: no like. 10% width on a mobile phone is like 20 px.

File: web/src/redux/githubID/githubID.reducer.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+//   notFound: boolean;
+// }
+
+// interface ProductOwnerGithubIDInterface extends GithubIdBaseInterface {
+//   productOwnerGithubID: {};
+// }
+// interface TechnicalLeadInitialState extends Array<GithubIdBaseInterface> {
+//   technicalLeads: {};
+// }
+// interface GithubIdInitialState extends ProductOwnerGithubIDInterface, TechnicalLeadInitialState {}
```
Comment (patricksimonian) [inline web/src/redux/githubID/githubID.reducer.ts:19]: whats this for

File: web/src/redux/githubID/githubID.reducer.ts [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+    notFound: false,
+  },
+  SecondUpdatedTechnicalLeads: {
+    githubUser: null,
+    inputKeyword: null,
+    isLoading: false,
+    everFetched: false,
+    notFound: false,
+  },
+};
```
Comment (patricksimonian) [inline web/src/redux/githubID/githubID.reducer.ts:43]: this could be better types. You could have a type for a 'user' for each type of persona.

**Review by patricksimonian — COMMENTED** [review]
A few comments, my biggest feedback is i'm not a huge fan of hardcoding in 'first' and 'second' technical leads. To me it appears they are both the same thing and could be better represented as an array ?

File: web/src/redux/githubID/githubID.reducer.ts (hunk 3 above)
Comment (w8896699) [reply web/src/redux/githubID/githubID.reducer.ts]: updated

File: web/src/redux/githubID/githubID.reducer.ts (hunk 2 above)
Comment (w8896699) [reply web/src/redux/githubID/githubID.reducer.ts]: now they are in use

File: web/src/components/common/UI/GithubUserValidation/AdaptedGithubUserDisplay.tsx [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+        <Text mt={2}>ID: {id}</Text>
+      </Flex>
+    </Flex>
+  );
+};
+
+const AdaptedGithubUserDisplay: React.FC<any> = (props) => {
+  const { input, githubIDAllState, persona, githubUserKeywordInputDispatch, ...rest } = props;
+
+  // TODO(Billy): For some reason, except for product owner, other two github id state is not been store in persistStore
```
Comment (patricksimonian) [inline web/src/components/common/UI/GithubUserValidation/AdaptedGithubUserDisplay.tsx:50]: is this TODO important for this PR? Could this be released in prod without this TODO?

File: web/src/components/common/UI/GithubUserValidation/AdaptedGithubUserDisplay.tsx [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+  useEffect(() => {
+    if (input.value !== githubIDAllState[persona].inputKeyword) {
+      githubUserKeywordInputDispatch(persona, input.value);
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [input]);
```
Comment (patricksimonian) [inline web/src/components/common/UI/GithubUserValidation/AdaptedGithubUserDisplay.tsx:57]: looks like this useEffect could be cleaned up. Also `input ` is a terrible variable name. 

```js
const { value } = input;

useEffect(() => {
    if (value !== githubIDAllState[persona].inputKeyword) {
     githubUserKeywordInputDispatch(persona, value) 
  }

}, [ value, persona, githubIDAllState, githubUserKeywordInputDispatch])

```

File: web/src/components/common/UI/GithubUserValidation/AdaptedGithubUserDisplay.tsx [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+};
+
+const AdaptedGithubUserDisplay: React.FC<any> = (props) => {
+  const { input, githubIDAllState, persona, githubUserKeywordInputDispatch, ...rest } = props;
+
+  // TODO(Billy): For some reason, except for product owner, other two github id state is not been store in persistStore
+
+  useEffect(() => {
+    if (input.value !== githubIDAllState[persona].inputKeyword) {
+      githubUserKeywordInputDispatch(persona, input.value);
```
Comment (patricksimonian) [inline web/src/components/common/UI/GithubUserValidation/AdaptedGithubUserDisplay.tsx:54]: this name isn't very descriptive also hard to understand. I'd suggest be prefixing the function with`dispatch` as in `dispatchGithubUserKeywordInput`

File: web/src/components/common/UI/GithubUserValidation/AdaptedGithubUserDisplay.tsx [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+      <Flex flexDirection="column" justifyContent="center">
+        <Text>Name: {name}</Text>
+        <Text mt={2}>ID: {id}</Text>
+      </Flex>
+    </Flex>
+  );
+};
+
+const AdaptedGithubUserDisplay: React.FC<any> = (props) => {
+  const { input, githubIDAllState, persona, githubUserKeywordInputDispatch, ...rest } = props;
```
Comment (patricksimonian) [inline web/src/components/common/UI/GithubUserValidation/AdaptedGithubUserDisplay.tsx:48]: githubIdAllState is also not a very good name. It doesn't really describe what it means

File: web/src/components/common/UI/GithubUserValidation/AdaptedGithubUserDisplay.tsx [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+      </Box>
+      <Flex flexDirection="column" justifyContent="center">
+        <Text>Name: {name}</Text>
+        <Text mt={2}>ID: {id}</Text>
+      </Flex>
+    </Flex>
+  );
+};
+
+const AdaptedGithubUserDisplay: React.FC<any> = (props) => {
```
Comment (patricksimonian) [inline web/src/components/common/UI/GithubUserValidation/AdaptedGithubUserDisplay.tsx:47]: This component could probably be made more `dumb` and less connected to `state`. 

Can't it receive all its props including a handler to dispatch the action?

File: web/src/components/common/UI/GithubUserValidation/GithubUserValidation.tsx [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+  const getPersona = (inputName: string) => {
+    const personaHandler = inputName.split(/[^A-Za-z0-9]/);
+    let persona = 'updatedProductOwner';
+    if (personaHandler[0].includes('TechnicalLeads')) {
+      if (personaHandler[1] === '0') {
+        persona = 'FirstUpdatedTechnicalLeads';
+      } else {
+        persona = 'SecondUpdatedTechnicalLeads';
+      }
+    }
```
Comment (patricksimonian) [inline web/src/components/common/UI/GithubUserValidation/GithubUserValidation.tsx:25]: what is this

Comment (patricksimonian) [reply web/src/components/common/UI/GithubUserValidation/GithubUserValidation.tsx:25]: no

File: web/src/components/common/UI/GithubUserValidation/GithubUserValidation.tsx [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  useEffect(() => {
+    const delayDebounceFn = setTimeout(() => {
+      // first condition: prevent first time render trigger api call because we already use peresis store.
+      // Second condition: only send api request if input change
+      if (githubUser?.login !== inputKeyword && inputKeyword.length !== 0) {
+        fetchUserStartAsync(inputKeyword, persona);
+      }
+    }, 1500);
+
+    return () => clearTimeout(delayDebounceFn);
```
Comment (patricksimonian) [inline web/src/components/common/UI/GithubUserValidation/GithubUserValidation.tsx:41]: good!

File: web/src/components/profileEdit/ContactCardEdit.tsx [hunk 11]
```diff
@@ -69,6 +82,17 @@ const ContactCardEdit: React.FC<IContactCardEditProps> = (props) => {
...
   const onSubmit = async (formData: any) => {
+    let onSubmitErrorMessage;
+    Object.entries(GithubIDAllState).forEach(([key, value]) => {
+      if (GithubIDAllState[key].everFetched === true && GithubIDAllState[key].notFound === true) {
+        onSubmitErrorMessage = 'Github User Not Found';
+      }
+      if (GithubIDAllState[key].inputKeyword && GithubIDAllState[key].everFetched === false) {
+        onSubmitErrorMessage = 'Loading Github User infomation';
+      }
+    });
```
Comment (patricksimonian) [inline web/src/components/profileEdit/ContactCardEdit.tsx:93]: this seems fishy

File: web/src/redux/githubID/githubID.action.ts [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+export const storeUser = (persona: string, payload: object) => ({
+  type: GithubIDActionTypes.GITHUB_USER_STORE_USER,
+  payload,
+  persona,
+});
+
+export const githubUserKeywordInput = (persona: Array<string>, payload: string) => ({
+  type: GithubIDActionTypes.GITHUB_USERS_INPUT,
+  payload,
+  persona,
```
Comment (patricksimonian) [inline web/src/redux/githubID/githubID.action.ts:26]: typically a redux action creator only has two top level fields `payload` and `type`. Everything else is within `payload`

File: web/src/utils/Wizard.tsx [hunk 13]
```diff
@@ -26,6 +31,20 @@ const Wizard: React.FC<any> = ({ onSubmit, children }) => {
...
+      let onSubmitErrorMessage;
+      Object.entries(GithubIDAllState).forEach(([key, value]) => {
+        if (GithubIDAllState[key].everFetched === true && GithubIDAllState[key].notFound === true) {
+          onSubmitErrorMessage = 'Github User Not Found';
+        }
+        if (GithubIDAllState[key].inputKeyword && GithubIDAllState[key].everFetched === false) {
+          onSubmitErrorMessage = 'Loading Github User infomation';
+        }
+      });
+      if (onSubmitErrorMessage) return { [FORM_ERROR]: onSubmitErrorMessage };
```
Comment (patricksimonian) [inline web/src/utils/Wizard.tsx:45]: duplicate function. equally fishhy

**Review by patricksimonian — CHANGES_REQUESTED** [review]
left some feedback we can go over

File: web/src/components/common/UI/GithubUserValidation/AdaptedGithubUserDisplay.tsx (hunk 4 above)
Comment (w8896699) [reply web/src/components/common/UI/GithubUserValidation/AdaptedGithubUserDisplay.tsx:50]: This TODO is a done, I forget to remove it.

File: web/src/redux/githubID/githubID.action.ts [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+import GithubIDActionTypes from './githubID.types';
+
+export type GithubIDAction = ActionType<typeof requestGithubUsers>;
+
+export const requestGithubUsers = (index: number) => ({
+  type: GithubIDActionTypes.GITHUB_USERS_REQUEST,
+  payload: index,
+});
+
+export const userExist = (index: number) => ({
```
Comment (patricksimonian) [inline web/src/redux/githubID/githubID.action.ts:12]: userExists

GITHUB USER EXISTS

File: web/src/utils/Wizard.tsx [hunk 15]
```diff
@@ -39,6 +39,11 @@ const Wizard: React.FC<any> = ({ onSubmit, children }) => {
       mutators={{
         ...arrayMutators,
       }}
+      validate={(eo) => {
+        const errors: any = {};
+
+        return errors;
+      }}
```
Comment (patricksimonian) [inline web/src/utils/Wizard.tsx:46]: this function does not use argument

File: web/src/redux/githubID/githubID.types.ts [hunk 16]
```diff
@@ -0,0 +1,10 @@
+const GithubIDActionTypes = {
+  GITHUB_USERS_REQUEST: 'GITHUB_USERS_REQUEST',
+  GITHUB_USERS_INPUT: 'GITHUB_USERS_INPUT',
+  GITHUB_USER_DOES_NOT_EXIST: 'GITHUB_USER_DOES_NOT_EXIST',
+  GITHUB_USER_EXIST: 'GITHUB_USER_EXIST',
```
Comment (patricksimonian) [inline web/src/redux/githubID/githubID.types.ts:5]: USER_EXISTS

File: web/src/utils/Wizard.tsx [hunk 17]
```diff
@@ -39,6 +39,11 @@ const Wizard: React.FC<any> = ({ onSubmit, children }) => {
       mutators={{
         ...arrayMutators,
       }}
+      validate={(eo) => {
```
Comment (patricksimonian) [inline web/src/utils/Wizard.tsx:42]:
```suggestion
      validate={(eo) => ({})}
```

File: web/src/components/common/UI/EditSubmitButton.tsx [hunk 18]
```diff
@@ -14,27 +14,31 @@
...
+    return (
+      <>
+        <StyledFormButton style={{ display: 'block' }}>Request Update</StyledFormButton>
+      </>
+    );
   }
   return (
     <>
-      {/* @ts-ignore */}
       <StyledFormDisabledButton disabled style={{ display: 'block' }}>
```
Comment (patricksimonian) [inline web/src/components/common/UI/EditSubmitButton.tsx:42]: no need to change but it is very interesting to change a button style from default `inline-block` to `block`. I'd personally just increase the width

File: web/src/components/profileEdit/ContactCardEdit.tsx [hunk 19]
```diff
@@ -172,100 +186,109 @@ const ContactCardEdit: React.FC<IContactCardEditProps> = (props) => {
...
+                        />
+                      </Flex>
+                      <Flex flexDirection="column">
+                        <Label htmlFor={`${name}.githubId`}>GitHub Id</Label>
+                        {/* Product Ownder has index 0, other TL will start from 1 */}
+                        <GithubUserValidation name={`${name}.githubId`} index={index + 1} />
+                      </Flex>
+                    </div>
+                  ))}
+                  {fields.length! < MAXIMUM_TECHNICAL_LEADS && (
```
Comment (patricksimonian) [inline web/src/components/profileEdit/ContactCardEdit.tsx:250]: i'd opt for `fields.length && fields.length < MAXIMUM_TECHNICAL_LEADS &&`

File: web/src/redux/githubID/githubID.action.ts (hunk 14 above)
Comment (patricksimonian) [inline web/src/redux/githubID/githubID.action.ts:12]: userExists

**Review by patricksimonian — APPROVED** [review]

**Pull request merged**